testsys: Refactor code base

ecpullen opened the pull request:
Changes the way TestSys crd's are created while keeping the same interface.

<!--
Tips:
- Please read CONTRIBUTING.md to understand our process and our requests for PRs.
- Please file an issue before creating a PR so we can discuss the change and confirm it's not already being worked on.
-->

**Issue number:**

Closes #2533 

**Description of changes:**

* The basis of the changes revolve around `CrdCreator`. The `CrdCreator` is responsible for creating all CRDs for a given variant family (`aws-k8s`, `aws-ecs`). The `CrdCreator` has a few functions that need to be implemented that enables support for the standard testing types. To add a new testing family, only a `CrdCreator` needs to be created.
* CRD creation is done based on structs instead of function parameters (`OrchestratorInput`, `BottlerocketInput`, etc).
* Additional labels are added, and support for conflicting resources has been streamlined using `CrdInput::existing_resources()`.
* EC2 CRD creation has been de-duped for `aws-k8s` and `aws-ecs`


**Testing done:**

Created ECS and EKS testing with:
```
cargo make -e BUILDSYS_VARIANT="aws-ecs-1" test
cargo make -e TESTSYS_TEST="migration" test
cargo make test
```

After testing was complete the following status was left:
```
 NAME                                TYPE           STATE           PASSED       SKIPPED      FAILED
 x86-64-aws-ecs-1-test               Test           passed          1            0            0
 x86-64-aws-k8s-123                  Resource       completed
 x86-64-aws-k8s-123-1-initial        Test           passed          1            7051         0
 x86-64-aws-k8s-123-2-migrate        Test           passed          2            0            0
 x86-64-aws-k8s-123-3-migrated       Test           passed          1            7051         0
 x86-64-aws-k8s-123-4-migrate        Test           passed          2            0            0
 x86-64-aws-k8s-123-5-final          Test           passed          1            7051         0
 x86-64-aws-k8s-123-test             Test           passed          1            7051         0
 ```
 
 Note: The instance providers are not in the status because `DestructionPolicy::OnTestSuccess` was used.

**Terms of contribution:**

By submitting this pull request, I agree that this contribution is dual-licensed under the terms of both the Apache License, version 2.0, and the MIT license.

ecpullen committed:
testsys: Refactor code base

Changes the way TestSys crd's are created while keeping the same
interface.

ecpullen commented:
^ Convert error from `anyhow` to `Snafu`.

ecpullen commented:
^ Add documentation

webern commented on tools/testsys/src/sonobuoy.rs:
```diff
@@ -0,0 +1,91 @@
+/// Create a Sonobuoy CRD for K8s conformance and quick testing.
+pub(crate) fn sonobuoy_crd(test_input: TestInput) -> Result<Test> {
+    let cluster_resource_name = test_input
+        .orchestrator_crd_name
```
I'm leaning toward using the word `cluster` instead of orchestrator since the word cluster is being used everywhere else already and introducing a new word is cognitively dissonant.

webern commented on tools/testsys/src/error.rs:
```diff
@@ -0,0 +1,70 @@
+    #[snafu(display("{} was missing from {}", item, what))]
+    Missing { item: String, what: String },
+
+    #[snafu(context(false), display("{}", source))]
```
Is context(false) because the Error type is not std::error::Error? I haven't used that feature before.
```
    #[snafu(context(false), display("{}", source))]
```

ecpullen replied:
`context(false)` makes it so you don't need to add the `.context()` for error conversion.

webern commented on tools/testsys/src/run.rs:
```diff
@@ -301,13 +263,16 @@ pub(crate) enum TestType {
+#[serde(untagged)]
+pub(crate) enum TestType {
+    Known(KnownTestType),
+    Unknown(String),
```
It's not clear looking at this definition what the `String` represents. As a doc comment for `Unknown`.

webern commented on tools/testsys/src/aws_resources.rs:
```diff
@@ -1,954 +1,39 @@
+use std::fs::File;
+
+/// Get the AMI for the given `region` from the `ami_input` file.
+pub(crate) fn ami(ami_input: &str, region: &str) -> Result<String> {
```
It feels like we should have a common struct for serializing/deserializing this

ecpullen replied:
There is but it is private in `Pubsys`

webern commented on tools/testsys/src/aws_resources.rs:
```diff
@@ -958,11 +43,14 @@ where
         .load()
         .await;
     let ec2_client = aws_sdk_ec2::Client::new(&config);
+    // Find all images named `name` on `arch` in the `region`.
```
The owners 'self' default assumes that this is being used against developer-built images, right? we can pass in the AMI-ID for other use cases?

ecpullen replied:
Correct.

webern commented on tools/testsys/src/aws_resources.rs:
```diff
@@ -984,12 +72,195 @@ where
 }
+
+#[derive(Clone, Debug, Deserialize)]
+pub(crate) struct AmiImage {
```
What was the purpose of this type? Validation?

ecpullen replied:
Easy deserialization. That was not added in the refactor.

etungsten reviewed (approved):
Refactor looks good to me. The testing looks good as well.

webern commented on tools/testsys/src/aws_ecs.rs:
```diff
@@ -0,0 +1,173 @@
+        get_ami_id(
+            format!(
+                "bottlerocket-{}-{}-{}-{}",
+                crd_input.variant, crd_input.arch, crd_input.starting_version.as_ref().context(error::InvalidSnafu{what: "The starting version must be provided for migration testing"})?, self.migrate_starting_commit.as_ref().context(error::InvalidSnafu{what: "The commit for the starting version must be provided if the starting image id is not"})?
```
Can you wrap this? Sadly rustfmt can't seem to handle things inside of macro calls.

webern commented on tools/testsys/src/aws_k8s.rs:
```diff
@@ -0,0 +1,155 @@
+        get_ami_id(
+            format!(
+                "bottlerocket-{}-{}-{}-{}",
+                crd_input.variant, crd_input.arch, crd_input.starting_version.as_ref().context(error::InvalidSnafu{what: "The starting version must be provided for migration testing"})?, self.migrate_starting_commit.as_ref().context(error::InvalidSnafu{what: "The commit for the starting version must be provided if the starting image id is not"})?
```
I'm seeing repeated code here. Maybe put this in an ami_name function.

webern commented on tools/testsys/src/aws_k8s.rs:
```diff
@@ -0,0 +1,155 @@
+            .crd_input
+            .existing_crds(
+                &labels,
+                &["testsys/cluster", "testsys/type", "testsys/region"],
```
Maybe we should think about constants for these. Technically even the array could be a constant but maybe that's not necessary.

ecpullen replied:
I think in this case, it's not necessary.

webern commented on tools/testsys/src/crds.rs:
```diff
@@ -0,0 +1,405 @@
+            &self.repo_config.metadata_base_url,
+            &self.repo_config.targets_url,
+        ) {
+            debug!(
```
Can we enable debug log level through `cargo make` currently? Aside: it might be nice to have a globally respected log level for Bottlerocket build and test processes that is respected throughout `Makefile.toml`.

ecpullen replied:
`-e RUST_LOG=debug` can be used to see those logs.

webern reviewed (approved):
Some recommendations. The biggest one is that I think we should stick with `cluster` over `orchestrator`, though I don't remember the reasoning behind the choice.

ecpullen commented:
^
* Changed `orchestrator` -> `cluster`
* Added doc comment for `TestType::Unknown`
* Manual text wrapping for large `format!` block
* Created `ami_name` function
